### PR TITLE
feat(kit): refactor Migrate to file-based workflow

### LINKED ISSUE #154: feat: refactor kit.Migrate to file-based workflow (DEVIATION:BROKEN)
## Context

\`kit.Migrate\` (and \`grizzle migrate\`) currently does **live introspect → diff → apply immediately**. This is \`DEVIATION:BROKEN\` relative to the Drizzle-parity target. See \`docs/spec/kit.md\` — *\"migrate — DEVIATION:BROKEN (current)\"*.

The target behaviour reads \`.sql\` files produced by \`grizzle generate\` and applies those not yet recorded in \`_grizzle_migrations\`.

## Current (broken) behaviour

```go
// Introspects live DB, diffs, applies — no files involved
result, err := kit.Migrate(ctx, pool, tables...)
```

## Target behaviour

```go
result, err := kit.Migrate(ctx, pool, kit.MigrateOptions{
    MigrationsDir: "./migrations",
})
```

Reads pending `.sql` files → applies in order → records in `_grizzle_migrations`.

## Blocked by

- #251 (transition plan spec — **approved**, PR #270)

## Transition plan

Fully documented in `docs/spec/kit.md` — see "History table transition plan" section. Key points:

- Target schema adds `tag TEXT NOT NULL DEFAULT ''` and `is_baseline BOOLEAN NOT NULL DEFAULT FALSE`
- Schema upgrade is automatic on first run (ADD COLUMN per dialect)
- Old rows preserved with `tag = ''`
- `--baseline <tag>` flag bootstraps file-based workflow on existing deployments
- Sequence numbers: four-digit zero-padded, leading integer before first underscore
- Tag format: `^[a-zA-Z0-9_-]+$`, all comparisons use parameterized queries

## Acceptance criteria

- [ ] `kit.Migrate` reads from migrations directory
- [ ] Applies pending files in sequence order
- [ ] Records each applied migration with `tag` (filename stem) and `is_baseline = FALSE`
- [ ] Calling migrate twice with unchanged schema is a no-op
- [ ] `_grizzle_migrations` table gains `tag` and `is_baseline` columns (all three dialects)
- [ ] Automatic schema upgrade on first run (ADD COLUMN per dialect, idempotent)
- [ ] `--baseline <tag>` marks migrations as applied without executing SQL; all inserts in a single transaction
- [ ] Tag format validated against `^[a-zA-Z0-9_-]+$` at library boundary
- [ ] All tag comparisons use parameterized queries
- [ ] `--skip-schema-upgrade` returns error if columns absent (no silent fallback)
- [ ] `Status()` and `LoadHistory()` updated to return `Tag` and `IsBaseline` in `MigrationRecord`
- [ ] Duplicate sequence prefixes rejected at both `generate` time and `migrate` time

## Reference

`docs/spec/kit.md` — migrate section and history table transition plan (PR #270)

### PR DESCRIPTION
**Pull request by charles-fineman**
## Summary

- Replaces the live-introspect → diff → apply approach in `kit.Migrate` / `MigrateMySQL` / `MigrateSQLite` with a file-based workflow that reads pending `.sql` files from a migrations directory and applies them in sequence order
- Adds `tag` and `is_baseline` columns to `_grizzle_migrations` for all three dialects with automatic schema upgrade on first run
- Implements `--baseline <tag>` flag to bootstrap existing deployments onto the file-based workflow without re-running already-applied SQL
- Tag format validated against `^[a-zA-Z0-9_-]+$` at the library boundary; all tag comparisons use parameterized queries
- Duplicate sequence prefixes rejected at migrate time

Closes #154

### Library API (`kit/`)

- `Migrate(ctx, pool, MigrateOptions{MigrationsDir})` — reads `.sql` files, applies pending ones in sequence order
- `MigrateMySQL(ctx, db, MigrateOptions)` / `MigrateSQLite(ctx, db, MigrateOptions)` — same for MySQL and SQLite
- `Status(ctx, pool, MigrateOptions)` — returns applied `MigrationRecord` rows and pending `MigrationFile` list
- `MigrationRecord` gains `Tag string` and `IsBaseline bool` fields
- `MigrateResult` gains `Applied []MigrationFile` and `Baselined []MigrationFile` (replaces `Changes []Change`)
- `LoadHistory` / `LoadHistoryMySQL` / `LoadHistorySQLite` return the updated struct
- New `migrate_files.go`: `LoadMigrationFiles`, `ValidateTag`, `ChecksumFile`, `MigrateOptions`
- `PushSQLite` / `DryRunSQLite` / `PushMySQL` / `DryRunMySQL` retain the live-introspect approach (push workflow is unchanged)

### CLI (`cmd/grizzle/main.go`)

- `grizzle migrate` now takes `--migrations <dir>` (replaces `--schema`) with `--baseline <tag>` and `--skip-schema-upgrade`
- `grizzle status` now takes `--migrations <dir>` with `--skip-schema-upgrade`

### Schema upgrade (all three dialects)

- Automatic `ADD COLUMN` for `tag` and `is_baseline` on first run against an old table
- Idempotent: guarded by `IF NOT EXISTS` (PostgreSQL) or `PRAGMA table_info` / `INFORMATION_SCHEMA` existence check (SQLite / MySQL)
- `--skip-schema-upgrade` returns an error if columns are absent (no silent fallback)

### Tests

- Rewrote `kit/kit_sqlite_test.go` SQLite integration tests for file-based workflow
- Added unit tests in `kit/kit_test.go` for `LoadMigrationFiles`, `ValidateTag`, duplicate prefix detection, ordering

## Test plan

- [x] `go test ./...` passes (all green)
- [x] `MigrateSQLite` with two migration files: both applied on first run, idempotent on second run
- [x] `--baseline` marks first migration as baseline, applies second normally
- [x] `--skip-schema-upgrade` returns error when columns absent
- [x] Schema upgrade idempotent (running on already-upgraded table has no effect)
- [x] Invalid tag format returns error before any DB operation
- [x] Duplicate sequence prefix returns error

### CODE REVIEW

**Comment by charles-fineman** [comment]
All peer reviews addressed. Ready for final review @scottescue

**Peer review summary:**
- Security: ✅ Approved — parameterized queries throughout, tag validation at library boundary, no injection vectors, proper audit logging to stderr for baseline operations
- Backend: ✅ Approved — clean MigrateOptions API, correct transaction boundaries, idempotent schema upgrade, proper error handling
- Testing: ✅ Approved — comprehensive SQLite integration tests covering all acceptance criteria
- Documentation: ✅ Approved — full spec update in kit.md including transition plan, all public API documented

**Deferred items captured as issues:**
- #273: Support dollar-quoted PL/pgSQL blocks in SQL splitter (P2 enhancement)
- #274: PostgreSQL integration tests for file-based kit.Migrate (P2 testing)

**Comment by charles-fineman** [comment]
All peer reviews addressed. Ready for final review @scottescue

**Summary of peer review changes:**
- Security: Added `table_schema = current_schema()` / `DATABASE()` filter to INFORMATION_SCHEMA queries in Postgres and MySQL to prevent false positives on shared servers (#272 security review)
- Backend: `Status()` now calls internal `loadHistory` (not public `LoadHistory`) to avoid double `ensureMigrationsTable` round-trips; `insertBaseline*` now accepts pre-computed checksums via `baselineRecord` to eliminate TOCTOU file re-reads; `baselineSeq` hoisted to avoid redundant `parseSequenceNumber` call; dead secondary sort key removed
- Docs: Updated "Current CLI" table and migrate section header (was still DEVIATION:BROKEN, now PARITY); PostgreSQL upgrade mechanism description clarified (unconditional idempotent SQL, not per-column detection); `LoadHistory` docstring corrected; `MigrateOptions.Baseline` comment updated with observable consequence; recommended upgrade path now includes caveat that `grizzle generate` is not yet implemented
- Tests: Added missing test for `Status()` non-zero Pending path, `is_baseline=FALSE` assertion on normal rows, multi-statement migration, `SkipSchemaUpgrade` success path, `LoadMigrationFiles` with non-existent dir, zero sequence prefix error, and more SeqNum assertions

**Follow-up issues created for deferred items:**
- #275: `splitSQLStatements` limitation with semicolons in string literals (P3)
- #276: MySQL/PostgreSQL integration tests for file-based workflow (P2)

File: kit/migrate.go [hunk 1]
```diff
@@ -21,157 +20,360 @@ const MigrationsTable = "_grizzle_migrations"
...
+
+	return StatusResult{Applied: history, Pending: pending}, nil
 }
 
 // LoadHistory returns all rows from _grizzle_migrations in chronological order.
 // Returns an empty slice (not an error) if the table does not exist yet.
 func LoadHistory(ctx context.Context, pool *pgxpool.Pool) ([]MigrationRecord, error) {
-	if err := ensureMigrationsTable(ctx, pool); err != nil {
+	if err := ensureMigrationsTable(ctx, pool, false); err != nil {
 		return nil, err
```
[the remark this answers is not in the document]
Comment (charles-fineman) [reply kit/migrate.go:234]: ❌ Already resolved. The current code in `Status` calls the internal `loadHistory` (not the public `LoadHistory`), so `ensureMigrationsTable` is only invoked once with `opts.SkipSchemaUpgrade` respected. This was addressed in commit `acbb95b` as part of peer review feedback on this PR. No additional change needed.

File: kit/migrate_sqlite.go [hunk 2]
```diff
@@ -71,102 +74,313 @@ func DryRunSQLite(ctx context.Context, db *sql.DB, tables ...pg.TableDefiner) (P
...
+	stmts := splitSQLStatements(sqlText)
 	for _, stmt := range stmts {
+		stmt = strings.TrimSpace(stmt)
+		if stmt == "" {
+			continue
+		}
 		// Skip comment-only statements (ALTER COLUMN stubs).
-		if strings.HasPrefix(strings.TrimSpace(stmt), "--") {
+		if strings.HasPrefix(stmt, "--") {
 			continue
```
[the remark this answers is not in the document]
Comment (charles-fineman) [reply kit/migrate_sqlite.go]: ✅ Fixed in ca086a3 and 247f3bf.

The root cause was that `applyMigrationFileSQLite` used `strings.HasPrefix(stmt, "--")` which discarded any semicolon-split fragment whose first line was a comment, even when real SQL followed.

Changes:
- Replaced the `HasPrefix` guard with a `stripLeadingCommentLines` helper that peels off leading blank and `--` comment lines, then skips only if nothing remains. ALTER COLUMN stubs (pure comment strings from the push workflow) are still correctly discarded.
- Added a clarifying comment to `execTransactionSQLite` explaining why it intentionally retains the simpler `HasPrefix` guard (it only handles DDL-generator output, never user-authored files).
- Added `TestStripLeadingCommentLines` (8 table-driven cases) and integration test `TestSQLite_MigrationWithLeadingComments_Executes`.

CI is green.

File: kit/migrate_mysql.go [hunk 3]
```diff
@@ -72,102 +75,319 @@ func DryRunMySQL(ctx context.Context, db *sql.DB, tables ...pg.TableDefiner) (Pu
...
+	placeholders := make([]string, len(cols))
+	args := make([]interface{}, len(cols)+1)
+	args[0] = MigrationsTable
+	for i, c := range cols {
+		placeholders[i] = "?"
+		args[i+1] = c
+	}
+	q := `SELECT column_name FROM information_schema.columns
+	      WHERE table_name = ?
+	        AND column_name IN (` + strings.Join(placeholders, ", ") + `)`
```
[the remark this answers is not in the document]
Comment (charles-fineman) [reply kit/migrate_mysql.go:300]: ❌ Already resolved. `columnExistenceMySQL` already includes `AND table_schema = DATABASE()` in its `information_schema.columns` query, scoping the check to the active database. This was added in commit `acbb95b` during peer review of this PR.

File: kit/migrate.go [hunk 4]
```diff
@@ -21,157 +20,374 @@ const MigrationsTable = "_grizzle_migrations"
...
 }
 
-// applyWithHistory runs the DDL statements and inserts a history record, all
-// in a single transaction so they succeed or fail together.
-func applyWithHistory(ctx context.Context, pool *pgxpool.Pool, stmts []string, checksum, desc string) error {
+// applyMigrationFilePostgres executes the SQL from a single migration file and
+// records it in the history table, all in one transaction.
+func applyMigrationFilePostgres(ctx context.Context, pool *pgxpool.Pool, tag, sql, checksum string) error {
 	tx, err := pool.Begin(ctx)
 	if err != nil {
```
[the remark this answers is not in the document]
Comment (charles-fineman) [reply kit/migrate.go:314]: 📋 Tracked in issue #277 — this touches non-transactional execution semantics (`CREATE INDEX CONCURRENTLY` etc.) which requires a new opt-in mechanism (e.g. a `-- grizzle:no-transaction` header convention or an autocommit detection path). That design is outside the scope of this PR's file-based workflow implementation.

File: kit/migrate.go [hunk 5]
```diff
@@ -21,157 +20,374 @@ const MigrationsTable = "_grizzle_migrations"
...
+				applied[rec.File.Tag] = true
+				fmt.Fprintf(os.Stderr, "grizzle: baseline %s  sha256:%s\n", rec.File.FileName, rec.Checksum)
+			}
+		}
+	}
+
+	// Apply pending files (seqNum > baselineSeq, or all if no baseline).
+	var toApply []MigrationFile
+	for _, f := range files {
+		if !applied[f.Tag] {
```
[the remark this answers is not in the document]
Comment (charles-fineman) [reply kit/migrate.go:157]: 📋 Tracked in issue #278 — checksum drift detection is a new guard feature (error vs. warning vs. `StatusResult` annotation is still a design decision) that goes beyond the file-based apply workflow this PR delivers. The PR's stated invariant is that files already recorded by tag are skipped; drift detection adds opt-in tamper/edit detection on top of that.

**Comment by charles-fineman** [comment]
✅ Auto-resolved merge conflicts with `main`.

(No human code owners listed in CODEOWNERS — `@sofired` is an organization account; please tag the right reviewer.)

**Resolved files:**
- `docs/spec/kit.md` — both sides changed this file from the same base. `main` restructured it into a thin spec index (adding dedicated spec files for normative content and explicitly stating "This file should not define legacy migration contracts or duplicate the normative file-migration specifications"). The `issue-154` branch updated the old inline detailed content to reflect the new file-based migrate workflow. Resolved by taking `main`'s thin-index structure: the detailed spec content added by this branch belongs in the dedicated spec files (`file-migrations-history.md`, `file-migrations-api.md`, etc.) already linked from the Spec Map.

**Verification:** `go build ./... && go vet ./... && go test -race -count=1 ./...` — all 13 packages passed.

Please re-review.

**Comment by scottescue** [comment]
The work in this PR as well as some already committed have strayed from project goals. I'm closing this PR while I work on improving the project specification and audit existing code.

File: kit/migrate.go [hunk 6]
```diff
@@ -249,3 +469,18 @@ func DescribeChanges(changes []Change) string {
 	}
 	return strings.Join(parts, "; ")
 }
+
+// splitSQLStatements naively splits a SQL text on ";" to produce individual
+// runnable statements. Empty statements (from trailing semicolons) are omitted.
+// This is intentionally simple: migration files are expected to contain
+// well-formed DDL, not PL/pgSQL dollar-quoted blocks.
+func splitSQLStatements(sql string) []string {
+	parts := strings.Split(sql, ";")
```
[the remark this answers is not in the document]
Comment (charles-fineman) [reply kit/migrate.go]: ✅ Fixed in e287f4f, adc86c3, 61fce4a.

Replaced `strings.Split(sql, ";") ` in `splitSQLStatements` with a character-level parser that skips semicolons inside:
- Single-quoted string literals (`'internal; do not drop'`) with `''` escape handling
- Double-quoted identifiers with `""` escape handling
- Backtick-quoted identifiers (MySQL)
- Line comments (`-- ...`)
- Block comments (`/* ... */`)

Dollar-quoted PL/pgSQL blocks remain out of scope per the existing comment; such migration files should not use this splitter.

Unterminated constructs (e.g. a malformed `'unclosed` literal) are documented: the partial content is flushed as a trailing statement and fails at SQL execution time with a clear database error. Error-return from the parser is tracked as a separate design decision in #280.

Also removed the redundant `strings.TrimSpace` + empty guard at the Postgres call site — `splitSQLStatements` already trims and filters.

14 table-driven tests cover all the new cases.
